### ISSUE TITLE
blobstore/lvolstore hits assert when loading after unclean shutdown

### ISSUE THREAD

**Issue by pawelkax** [spdk/spdk#291]
## Expected Behavior
Spdk_tgt starts normally.

## Current Behavior
Sometimes following error occurs:
spdk_tgt: blobstore.c:81: _spdk_bs_claim_cluster: Assertion `spdk_bit_array_get(bs->used_clusters, cluster_num) == false' failed. 
After that vhost crashes.

## Possible Solution
None

## Steps to Reproduce
1. Start spdk_tgt (config contains two nvme definition) 
2. Split nvme disc into two bdevs.
3. Create lvol_store on first split.
4. Create thin and thick provision lvol bdev.
5. Create snapshot of thick provision lvol bdev.
6. Create clone of snapshot.
6. Kill spdk_tgt.
7. Try to start spdk_tgt again.

## Context (Environment including OS version, SPDK version, etc.)
Virtual machine with ubuntu and two emulated nvme devices (1G, 2G).

**Comment by dverkamp-intel** [comment]
Based on the assert, this looks like a blobstore or lvol issue, probably not related to vhost.

**Comment by jimharris** [comment]
Can you provide a few more specifics?


1)       How many splits of the NVMe bdev?

2)       Do you create an lvolstore on each split?

3)       Do you create a thin and think provision lvol bdev on each lvolstore?

4)       Which lvol do you create a snapshot of?  Same for the clone?

Thanks,

-Jim

**Comment by dverkamp-intel** [comment]
Here's a simple script that reproduces the issue every time for me (run from the root of a git checkout of spdk):

~~~
#!/usr/bin/env bash

source test/common/autotest_common.sh

cat > /tmp/spdk.conf <<EOF
[Ioat]
  Disable Yes
EOF

./scripts/gen_nvme.sh | head -2 >> /tmp/spdk.conf

app/spdk_tgt/spdk_tgt -c /tmp/spdk.conf &
pid=$!

waitforlisten $pid

./scripts/rpc.py destroy_lvol_store -l Lvs0 || true # clean up from previous test runs
./scripts/rpc.py construct_lvol_store Nvme0n1 Lvs0
./scripts/rpc.py construct_lvol_bdev -l Lvs0 -t Lvol0 100

# send SIGKILL to prevent spdk_tgt from doing a clean shutdown
kill -9 $pid
wait $pid || true

# start up the target again
app/spdk_tgt/spdk_tgt -c /tmp/spdk.conf &
pid=$!

# assert will fire here:
# spdk_tgt: blobstore.c:81: _spdk_bs_claim_cluster: Assertion `spdk_bit_array_get(bs->used_clusters, cluster_num) == false' failed.

waitforlisten $pid

# if we get here, everything worked; clean up
./scripts/rpc.py destroy_lvol_store -l Lvs0

kill $pid
wait $pid

echo "done"
~~~

Note that after you run this, you'll be left with a broken lvolstore that hits the assert every time, so you will probably need to attach that device to the kernel nvme driver and clear the superblock before you can test again, e.g.:

~~~
scripts/setup.sh reset
dd if=/dev/zero of=/dev/nvme0n1 bs=1M count=1 oflag=direct
~~~

**Comment by jimharris** [comment]
Thanks – I’m taking a look now.

**Comment by jimharris** [comment]
Proposed fix plus unit test here:  https://review.gerrithub.io/#/c/spdk/spdk/+/410065/

**Comment by jimharris** [comment]
@pawelkax - a fix was merged to master for this issue.  Can you please confirm it fixes your test case?

Thanks,

-Jim

**Comment by pawelkax** [comment]
I can confirm that it fixes the problem.

**Comment by dverkamp-intel** [comment]
Closing this as fixed in commit e8ddb060f87252c596adbd4ccb492df7e2f7925d
